setup.py

```diff
@@ -658,9 +655,10 @@ def write_version(filename: str = os.path.join(*[my_dir, "airflow", "git_version
 
 EXTRAS_PROVIDERS_PACKAGES: Dict[str, Iterable[str]] = {
     'all': list(PROVIDERS_REQUIREMENTS.keys()),
-    # this is not 100% accurate with devel_ci definition, but we really want to have all providers
-    # when devel_ci extra is installed!
+    # this is not 100% accurate with devel_ci and devel_all definition, but we really want
+    # to have all providers when devel_ci extra is installed!
     'devel_ci': list(PROVIDERS_REQUIREMENTS.keys()),
+    'devel_all': list(PROVIDERS_REQUIREMENTS.keys()),
```

Comment:
What's the difference between devel_ci and devel_all now?

Reply:
It appears the only difference is that devel_all includes beam.

In this case I'd be tempted to say that we don't need a new extra, and we can stick with `devel_ci` (or rename that to `devel_all`) -- but having two groups isn't necessary.

Reply:
My understanding is that `devel_all` is everything you may need to develop Airflow (100%). Then `develop_ci` is everything we need to run tests and avoid dependency problems when generating constraints - beam is a big pain. CC @potiuk who may know more

Reply:
The only place beam is used in inside `airflow/providers/google/cloud/utils/mlengine_prediction_summary.py` -- and from what I can tell, that isn't actually _run_ on Airflow machines, but uploaded to some GCP service?

I.e. beam is never a dep of Airflow? Am I skimming the code right?

Reply:
CC @TobKed @mik-laj who are more familiar with beam in Airflow (mlengine and dataflow)

Reply:
The `apache.beam` is an optional dependency that is needed for Dataflow  https://github.com/apache/airflow/blob/master/airflow/providers/google/cloud/operators/dataflow.py.

I think @mik-laj worked mostly on this (but it has been there for quite some time before @mik-laj improved it and made a truly optional requirement). 

As far as I know, previously "beam" requirements were implicitly needed when you wanted to run the Dataflow operator. But @mik-laj changed it in the way that you can optionally install some additional packages (including your own version of beam package if you need). in a virtualenv dynamically created when the task is running. 

So technically speaking beam is not needed for dataflow, but the additional virtualenv has many drawbacks - it takes quite  time to install, in many environments (including composer) you are not allowed to install additional packages on-the-flight and you can have very brittle environment - for example you can have conflicts when installing this new packages. Therefore the possibility of using pre-installed apache-beam (as airflow extra) is also retained - in this case the extra is useful if you want to install for example only google provider and use only that (so that the many dependencies of apache.beam do not conflict with dependencies of other extras).

That's why beam has been excluded from devel_ci - because we are running CI tests with the assumption that someone uses extra virtualenv to run the Python Dataflow operator.

I am also using this extra to install everything needed when I verify backport operators. During this check, I import all classes from all providers and if we have no "apache.beam" installed - some of the classes cannot be imported.

Here is the relevant line from the Dataflow operator: https://github.com/apache/airflow/blob/master/airflow/providers/google/cloud/operators/dataflow.py#L796

I hope it explains it all? And of course we can propose another solution for that, maybe simply `devel_plus_beam` for example should be a good name for the extra rather than `devel_all`. That correctly reflects the expectation.

Reply:
If beam is truly a dep of airflow (or an optional dep of a dep) then `devel_all` makes sense and it can stay as it is -- I just wasn't able to see anywhere in Airflow code that actually did anything with beam other than this file

https://github.com/apache/airflow/blob/c34ef853c890e08f5468183c03dc8f3f3ce84af2/airflow/providers/google/cloud/utils/mlengine_prediction_summary.py

Which is used via

https://github.com/apache/airflow/blob/c34ef853c890e08f5468183c03dc8f3f3ce84af2/airflow/providers/google/cloud/utils/mlengine_operator_utils.py#L242-L254

i.e. not on Airflow, but on a remote datprocjob.

And the test doesn't _need_ beam, it could be re-worked to mock the import I feel https://github.com/apache/airflow/blob/7c6dfcb0bfa2d7654996ae8e3d77ea5a0719089d/tests/providers/google/cloud/utils/test_mlengine_prediction_summary.py. Happy to tackle this myself it people think it's a good idea.

Reply:
Apache Beam is a big project and has/has a lot of dependency conflicts, especially older versions. For example: Apache Beam v2.21 required a mock older than 3, which was in conflict with our tests.
https://github.com/apache/beam/blob/v2.21.0/sdks/python/setup.py#L157

It is also worth adding that there are several other ways to run Dataflow jobs. Instead of `DataflowCreatePythonJobOperator` and running this process locally, we can use `DataflowTemplatedJobStartOperator` or `DataflowStartFlexTemplateOperator` or `KubernetesPodOperator`. 
I will be dissuaded from running Apache Beam on one cluster with Airflow, because it is often more problematic and less optimal (also financially). I have high hopes for `DataflowStartFlexTemplateOperator`, which was implemented by @TobKed  and addresses most of the use cases that users have had.

CC: @aaltay

Reply:
From my experience installing both beam and airflow (including 2/0) requires some additional constrains (thus #12508 ). @ashb should we be able to move on with the above explanation from @potiuk and @mik-laj ?

Reply:
Not quite yet -- no one has yet told me where _in Airflow_ we actually use Beam code. When do we import it?

Reply:
We don't import, but run Python with Beam as a subprocess. By default, this is the same interpreter as Airflow.
https://github.com/apache/airflow/blob/936566c586e6cbb155ffa541e89a31f7239f51bb/airflow/providers/google/cloud/operators/dataflow.py#L755
https://github.com/apache/airflow/blob/936566c586e6cbb155ffa541e89a31f7239f51bb/airflow/providers/google/cloud/operators/dataflow.py#L941
https://github.com/apache/airflow/blob/936566c586e6cbb155ffa541e89a31f7239f51bb/airflow/providers/google/cloud/hooks/dataflow.py#L818
https://github.com/apache/airflow/blob/936566c586e6cbb155ffa541e89a31f7239f51bb/airflow/providers/google/cloud/hooks/dataflow.py#L894
https://github.com/apache/airflow/blob/936566c586e6cbb155ffa541e89a31f7239f51bb/airflow/providers/google/cloud/hooks/dataflow.py#L569-L570
https://github.com/apache/airflow/blob/936566c586e6cbb155ffa541e89a31f7239f51bb/airflow/providers/google/cloud/hooks/dataflow.py#L437
It works similar to a pyspark.
https://github.com/apache/airflow/blob/936566c586e6cbb155ffa541e89a31f7239f51bb/airflow/providers/apache/spark/hooks/spark_submit.py#L418
https://github.com/apache/airflow/blob/936566c586e6cbb155ffa541e89a31f7239f51bb/setup.py#L388

Reply:
In system testing, we define Apache Beam versions, which means a virtual environment is created.
https://github.com/apache/airflow/blob/936566c586e6cbb155ffa541e89a31f7239f51bb/airflow/providers/google/cloud/hooks/dataflow.py#L886
https://github.com/apache/airflow/blob/936566c586e6cbb155ffa541e89a31f7239f51bb/airflow/providers/google/cloud/example_dags/example_dataflow.py#L115

Reply:
Gotcha! Sorry, was totally mis-understanding the code. And thanks for the very detailed explanation!

Reply:
Folks let me know if I can answer a question here. It looks like the questions are answered already :)